docs/source/etcd-architecture.rst

```diff
@@ -0,0 +1,238 @@
+particular, it reports errors and problems with configuring its host. This data
+is written into etcd, to make it visible to other components and operators of
+the network.
+
+
+.. _calico-orchestrator-plugin:
+
+Orchestrator Plugin
+-------------------
+
+Unlike Felix there is no single 'orchestrator plugin': instead, there are
+separate plugins for each major cloud orchestration platform (e.g. OpenStack,
+Kubernetes). The purpose of these plugins is to bind Calico more tightly into
+the orchestrator, allowing users to manage the Calico network just as they'd
+manage network tools that were built in to the orchestrator.
+
+A good example of an orchestrator plugin is the Calico Neutron ML2 mechanism
+driver. This component integrates with Neutron's ML2 plugin, and allows users
+to configure the Calico network by making Neutron API calls. This provides
+seamless integration with Neutron, and lowers the barrier of entry to using
+Calico.
+
+The orchestrator plugin is repsonsible for the following tasks:
+
+API Translation
+~~~~~~~~~~~~~~~
+
+The orchestrator will inevitably have its own set of APIs for managing
+networks. The orchestrator plugin's primary job is to translate those APIs into
+the Calico etcd data model (see :doc:`etcd-data-model`) to allow Calico to
+perform the appropriate network programming.
+
+Some of this translation will be very simple, other bits may be more complex in
+order to render a single complex operation (e.g. live migration) into the
+series of simpler operations the rest of the Calico network expects.
+
+Feedback
+~~~~~~~~
+
+If necessary, the orchestrator plugin will provide feedback from the Calico
+network into the orchestrator. Examples include: providing information about
+Felix liveness; marking certain endpoints as failed if network setup failed.
+
+
+.. _calico-etcd-component:
+
+etcd
+----
+
+Calico uses etcd as its backing data store. etcd is a distributed key-value
+store that has a focus on consistency, ensuring that Calico can always build an
+accurate network.
+
+In addition to its role as Calico's primary data store, etcd also acts as a
+communication mechanism between the various components. We do this by having
+the non-etcd components watch certain points in the keyspace to ensure that
+they see any changes that have been made, allowing them to respond to those
+changes in a timely manner.
+
+The etcd component is distributed across the entire deployment. It is divided
+into two groups of machines: the core cluster, and the proxies.
+
+In Calico, we deploy an etcd cluster. In small deployments this can be an etcd
+cluster of one, which provides no redundancy but is simple and low-overhead.
+In larger deployments we scale this up, as per the `etcd admin guide`_.
+
+Additionally, on each machine that hosts either a
+:ref:`calico-felix-component` or a :ref:`calico-orchestrator-plugin`, we run
+an etcd proxy. This is an attempt to reduce load on the core cluster and to
```

Comment:
OK - my suggested text above doesn't make sense if the plug-in has a proxy as well even in the simple case...